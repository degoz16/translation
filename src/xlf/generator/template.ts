import languages from '@cospired/i18n-iso-languages';
import countries from '@shellscape/i18n-iso-countries';

export type TemplateParameters = {
    source: LanguageLocale;
    target: LanguageLocale;
    skeletonPath: string;
    markdownPath: string;
};

export type LanguageLocale = {
    language: Language;
    locale: countries.Alpha2Code;
};

const languagesList = languages.langs();

const gap = (ident: number) => (chunk: string) => {
    const gap = ' '.repeat(ident);
    return chunk.split('\n').map((part) => gap + part).join('\n')
};

export type Language = (typeof languagesList)[number];

function generateTemplate(parameters: TemplateParameters, units: string[]) {
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }
    const {source, target, skeletonPath, markdownPath} = parameters;
    let indentation = 0;
    let before = '<?xml version="1.0" encoding="UTF-8"?>';
    before += '\n';

    before += '<xliff xmlns="urn:oasis:names:tc:xliff:document:1.2" version="1.2">';
    before += '\n';

    indentation += 2;

    before += ' '.repeat(indentation);
    before += `<file original="${markdownPath}" source-language="${source.language}-${source.locale}" target-language="${target.language}-${target.locale}" datatype="markdown">`;
    before += '\n';

    indentation += 2;

    before += ' '.repeat(indentation);
    before += '<header>';
    before += '\n';

    indentation += 2;

    before += ' '.repeat(indentation);
    before += '<skeleton>';
    before += '\n';

    indentation += 2;

    before += ' '.repeat(indentation);
    before += `<external-file href="${skeletonPath}"></external-file>`;
    before += '\n';

    indentation -= 2;

    before += ' '.repeat(indentation);
    before += '</skeleton>';
    before += '\n';

    indentation -= 2;

    before += ' '.repeat(indentation);
    before += '</header>';
    before += '\n';

    before += ' '.repeat(indentation);
    before += '<body>';
    before += '\n';

    before += units.map(gap(indentation)).join('\n');
    before += '\n';

    let after = ' '.repeat(indentation);
    after += '</body>';
    after += '\n';

    indentation -= 2;

    after += ' '.repeat(indentation);
    after += '</file>';
    after += '\n';

    indentation -= 2;
    after += ' '.repeat(indentation);
    after += '</xliff>';
    after += '\n';

    return before + after;
}

function validParameters(parameters: TemplateParameters) {
    const {source, target, skeletonPath, markdownPath} = parameters;

    const conditions = [
        ...[source, target].map(validLanguageLocale),
        ...[skeletonPath, markdownPath].map(Boolean),
    ];

    return conditions.reduce((a, v) => a && v, true);
}

function validLanguageLocale(parameters: LanguageLocale) {
    const {language, locale} = parameters;

    return languages.isValid(language) && countries.isValid(locale);
}

export {generateTemplate, validParameters};
export default {generateTemplate, validParameters};
