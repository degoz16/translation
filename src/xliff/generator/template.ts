import {ok} from 'assert';
import languages from '@cospired/i18n-iso-languages';
import countries from '@shellscape/i18n-iso-countries';

export type TemplateParams = {
    source: LanguageLocale;
    target: LanguageLocale;
    skeletonPath?: string;
    markdownPath?: string;
};

export type LanguageLocale = {
    language: Language;
    locale: countries.Alpha2Code;
};

const languagesList = languages.langs();

export type Language = (typeof languagesList)[number];

function unit(source: string, index: number) {
    return `
      <trans-unit id="${index + 1}">
        ${source}
      </trans-unit>    
    `.trim();
}

export function generate(parameters: TemplateParams, units: string[]) {
    validateParams(parameters);

    const {source, target, skeletonPath, markdownPath} = parameters;
    return `
<?xml version="1.0" encoding="UTF-8"?>
<xliff xmlns="urn:oasis:names:tc:xliff:document:1.2" version="1.2">
  <file original="${markdownPath}" source-language="${source.language}-${
      source.locale
  }" target-language="${target.language}-${target.locale}" datatype="markdown">
    <header>
      <skeleton>
        <external-file href="${skeletonPath}"></external-file>
      </skeleton>
    </header>
    <body>
      ${units.map(unit).join('\n      ')}
    </body>
  </file>
</xliff>
    `.trim();
}

function validateParams(parameters: TemplateParams) {
    const {source, target} = parameters;

    ok(validLanguageLocale(source), 'Invalid source language locale pair.');
    ok(validLanguageLocale(target), 'Invalid target language locale pair.');
}

function validLanguageLocale(parameters: LanguageLocale) {
    const {language, locale} = parameters;

    return languages.isValid(language) && countries.isValid(locale);
}
