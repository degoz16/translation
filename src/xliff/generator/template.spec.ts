import {template} from './template';

describe('smoke', () => {
  it('works', () => {
    const parameters = {
      file: {
        path: '/file.md',
        data: '',
      },
      source: {
        language: 'ru',
        locale: 'RU' as const,
      },
      target: {
        language: 'en',
        locale: 'US' as const,
      },
    };

    template([], parameters);
  });
});

describe('template', () => {
  it('generates xliff template', () => {
    const parameters = {
      file: {
        path: 'file.md',
        data: '',
      },
      source: {
        language: 'ru',
        locale: 'RU' as const,
      },
      target: {
        language: 'en',
        locale: 'US' as const,
      },
    };

    expect(template([], parameters)).toMatchSnapshot();
  });
});
