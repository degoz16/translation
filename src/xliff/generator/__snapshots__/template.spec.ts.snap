// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template generates xliff template 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<xliff xmlns=\\"urn:oasis:names:tc:xliff:document:1.2\\" version=\\"1.2\\">
  <file original=\\"file.ext\\" source-language=\\"ru-RU\\" target-language=\\"en-US\\" datatype=\\"markdown\\">
    <header>
      <skeleton>
        <external-file href=\\"file.skl\\"></external-file>
      </skeleton>
    </header>
    <body>
      
    </body>
  </file>
</xliff>"
`;
